src/IO/S3/Requests.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <IO/S3/URI.h>
+
+#include <aws/core/endpoint/EndpointParameter.h>
+#include <aws/s3/model/HeadObjectRequest.h>
+#include <aws/s3/model/ListObjectsV2Request.h>
+#include <aws/s3/model/ListObjectsRequest.h>
+#include <aws/s3/model/GetObjectRequest.h>
+#include <aws/s3/model/AbortMultipartUploadRequest.h>
+#include <aws/s3/model/CreateMultipartUploadRequest.h>
+#include <aws/s3/model/CompleteMultipartUploadRequest.h>
+#include <aws/s3/model/CopyObjectRequest.h>
+#include <aws/s3/model/PutObjectRequest.h>
+#include <aws/s3/model/UploadPartRequest.h>
+#include <aws/s3/model/UploadPartCopyRequest.h>
+#include <aws/s3/model/DeleteObjectRequest.h>
+#include <aws/s3/model/DeleteObjectsRequest.h>
+
+namespace DB::S3
+{
+
+namespace Model = Aws::S3::Model;
+
+template <typename BaseRequest>
+class ExtendedRequest : public BaseRequest
+{
+public:
+    Aws::Endpoint::EndpointParameters GetEndpointContextParams() const override
+    {
+        auto params = BaseRequest::GetEndpointContextParams();
+        if (!region_override.empty())
+            params.emplace_back("Region", region_override);
+
+        if (uri_override.has_value())
+        {
+            static const Aws::String AWS_S3_FORCE_PATH_STYLE = "ForcePathStyle";
+            params.emplace_back(AWS_S3_FORCE_PATH_STYLE, !uri_override->is_virtual_hosted_style);
+            params.emplace_back("Endpoint", uri_override->endpoint);
+        }
+
+        return params;
+    }
+
+    void overrideRegion(std::string region) const
+    {
+        region_override = std::move(region);
+    }
+
+    void overrideURI(S3::URI uri) const
+    {
+        uri_override = std::move(uri);
+    }
+
+    const auto & getURIOverride() const
+    {
+        return uri_override;
+    }
+
+protected:
+    mutable std::string region_override;
+    mutable std::optional<S3::URI> uri_override;
+};
+
+using HeadObjectRequest = ExtendedRequest<Model::HeadObjectRequest>;
```

Comment:
Can't we just modify `BaseRequest` without inheriting every S3 request type?

Reply:
I wanted to avoid modifying the SDK because it's more complex to extend with new types (update fork then update the commit of the submodule). This way it's much easier to add support for new type of request IMO.